cloudinit/lifecycle.py

```diff
@@ -0,0 +1,237 @@
+
+    @classmethod
+    def from_str(cls, version: str) -> "Version":
+        """Create a Version object from a string.
+
+        :param version: A period-delimited version string, max 4 segments.
+
+        :raises TypeError: Raised if invalid arguments are given.
+        :raises ValueError: Raised if invalid arguments are given.
+
+        :return: A Version object.
+        """
+        return cls(*(list(map(int, version.split(".")))))
+
+    def __gt__(self, other):
+        return 1 == self._compare_version(other)
+
+    def __eq__(self, other):
+        return (
+            self.major == other.major
+            and self.minor == other.minor
+            and self.patch == other.patch
+            and self.rev == other.rev
+        )
+
+    def __iter__(self):
+        """Iterate over the version (drop sentinels)"""
+        for n in (self.major, self.minor, self.patch, self.rev):
+            if n != -1:
+                yield str(n)
+            else:
+                break
+
+    def __str__(self):
+        return ".".join(self)
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def _compare_version(self, other: "Version") -> int:
+        """Compare this Version to another.
+
+        :param other: A Version object.
+
+        :return: -1 if self > other, 1 if self < other, else 0
+        """
+        if self == other:
+            return 0
+        if self.major > other.major:
+            return 1
+        if self.minor > other.minor:
+            return 1
+        if self.patch > other.patch:
+            return 1
+        if self.rev > other.rev:
+            return 1
+        return -1
+
+
+def should_log_deprecation(version: str, boundary_version: str) -> bool:
+    """Determine if a deprecation message should be logged.
+
+    :param version: The version in which the thing was deprecated.
+    :param boundary_version: The version at which deprecation level is logged.
+
+    :return: True if the message should be logged, else False.
+    """
+    return boundary_version == "devel" or Version.from_str(
+        version
+    ) <= Version.from_str(boundary_version)
+
+
+def log_if_acceptable(
+    *, logger: logging.Logger, version: str, level: int, msg: str, args
+):
+    """Log a message at the requested level, if that is acceptable.
+
+    If the log level is too high due to the version boundary, log at DEBUG
+    level. Useful to add new warnings to previously unguarded code without
+    disrupting stable downstreams.
+
+    :param logger: Logger object to log with
+    :param version: Version string of the version that this log was introduced
+    :param level: Preferred level at which this message should be logged
+    :param msg: Message, as passed to the logger.
+    :param args: Message parameters
+
+    :return: True if the message should be logged, else False.
+    """
+    if should_log_deprecation(version, features.DEPRECATION_INFO_BOUNDARY):
```

Comment:
If I were to call `log_if_acceptable()` at level INFO, then it would downgrade it to DEBUG? I'm guessing there's no good reason to do that, so I don't think this needs to change, but I just want to make sure we don't think we need to do any `level` checking here first.

Reply:
> If I were to call log_if_acceptable() at level INFO, then it would downgrade it to DEBUG?

Correct

> I'm guessing there's no good reason to do that, so I don't think this needs to change, but I just want to make sure we don't think we need to do any level checking here first.

Thanks for raising this point. It wouldn't make sense to use it in that way, but this is just a simple helper intended to make this logic less error-prone and callsites more readable. I'm not too worried about misuse because the docstring explains its intent, and this should be used on a case-by-case basis, and also the consequences of misuse are pretty minor. I agree with leaving it as is without level checking.